xla/hlo/ir/backend_config.cc

```diff
@@ -57,6 +60,13 @@ const std::string& BackendConfigWrapper::GetRawStringWithoutMutex() const {
   return raw_string_.empty() ? *kEmptyString : raw_string_;
 }
 
+absl::StatusOr<std::string> BackendConfigWrapper::GetFingerprint() const {
+  absl::MutexLock lock{&mutex_};
+  std::string result;
+  TF_RET_CHECK(tsl::SerializeToStringDeterministic(*proto_, &result));
+  return absl::BytesToHexString(result);
```

Comment:
BytesToHexString is not just a fingerprint (hash), it's a full, reversible encoding of the original data.

Can you explain what the use-case is? You might as well just return the result from `SerializeToStringDeterministic` or event the proto directly, at which point I'm not even sure we need this method.

Reply:
>BytesToHexString is not just a fingerprint (hash), it's a full, reversible encoding of the original data.

A fingerprint does not have to be a hash. The current one may be a little more practical if someone wanted to load it back from an HLO dump.

> Can you explain what the use-case is?

We need structures like map<hlo_instruction, compiled_binary> where the binary also depends on the instruction's backend config.

> You might as well just return the result from SerializeToStringDeterministic

Yes, though that would be harder to use in debugging, because that contains non-printable binary sequences.
The current output of GetFingerprint() can for example be used in HLO dumps or in logging.

>or event the proto directly

What do you mean? We need some deterministic string representation which can be attached to the string representation of everything else in the HLO instruction.

Reply:
The HLO instruction can already print the JSON version of the backend config. This can be used for a full-round trip if necessary. From the text above I don't get the feeling that such roundtripping is essential to your use case.

Calling the current solution in this PR a "fingerprint" is misleading, because a fingerprint is commonly understood to be a just an "identifier", not a complete encoding of the full data. If we proceed with this PR as is, the HLO instruction dump would have two different ways to express the entire backend config, which is not what we want.

Is the main missing piece here the determinism of the encoding? Is the JSON encoding non-deterministic? If so, then we should rather find a way to make it deterministic, rather than introducing an extra encoding.

Reply:
>a fingerprint is commonly understood to be a just an "identifier"

See https://github.com/openxla/xla/blob/a2e537951445a1d558ddc31a3ae33c829c40c52c/xla/service/gpu/kernel_reuse_cache.cc#L70 for counterexample. Though I don't mind using a hash or renaming the method.

> the HLO instruction dump would have two different ways to express the entire backend config

There are many ways to print HLO: https://github.com/openxla/xla/blob/d8f3bc8212c6074c716c5fc57cbb918393ac92b8/xla/hlo/ir/hlo_instruction.h#L92

> Is the main missing piece here the determinism of the encoding? Is the JSON encoding non-deterministic?

Yes, see the description of the PR.

> we should rather find a way to make it deterministic

That would be nice, do you see a reasonable way to do so?